src/sentry/monitors/consumers/monitor_consumer.py

```diff
@@ -566,30 +566,75 @@ def _process_checkin(
-) -> None:
-    try:
-        try_monitor_tasks_trigger(ts, partition)
-    except Exception:
-        logger.exception("Failed to trigger monitor tasks", exc_info=True)
+_checkin_workwer = ThreadPoolExecutor()
 
-    # Nothing else to do with clock pulses
-    if wrapper["message_type"] == "clock_pulse":
-        return
 
-    with sentry_sdk.start_transaction(
-        op="_process_message",
-        name="monitors.monitor_consumer",
-    ) as txn:
-        params: CheckinPayload = json.loads(wrapper["payload"])
-        start_time = to_datetime(float(wrapper["start_time"]))
-        project_id = int(wrapper["project_id"])
-        source_sdk = wrapper["sdk"]
+def process_checkin_group(items: List[CheckinItem]):
+    for item in items:
+        try:
+            with sentry_sdk.start_transaction(
+                op="_process_checkin",
+                name="monitors.monitor_consumer",
+            ) as txn:
+                _process_checkin(item, txn)
+        except Exception:
+            logger.exception("Failed to process check-in", exc_info=True)
+            continue
+
+
+def process_batch(message: Message[ValuesBatch[KafkaPayload]]):
+    """
+    Receives batches of check-in messages. This function will take the batch
+    and group them together by monitor ID (ensuring order is preserved) and
+    execute each group using a ThreadPoolWorker.
+
+    By batching we're able to process check-ins in paralell while guaranteeing
+    that no check-ins are processed out of order per monitor environment.
+    """
+    batch = message.payload
+
+    latest_partition_ts: Mapping[int, datetime] = {}
+    checkin_mapping: Mapping[str, List[CheckinItem]] = defaultdict(list)
+
+    for item in batch:
+        assert isinstance(item, BrokerValue)
+
+        try:
+            wrapper: CheckinMessage | ClockPulseMessage = msgpack.unpackb(item.payload.value)
+        except Exception:
+            logger.exception("Failed to unpack message payload", exc_info=True)
+            continue
 
-        _process_checkin(params, ts, start_time, project_id, source_sdk, txn)
+        latest_partition_ts[item.partition.index] = item.timestamp
+
+        # Nothing needs to be done with a clock pulse, we will have already
+        # stored the latest_partition_ts to be used to tick the clock at the
+        # end of this batch if necessary
+        if wrapper["message_type"] == "clock_pulse":
+            continue
+
+        item = CheckinItem(
+            ts=item.timestamp,
+            partition=item.partition.index,
+            message=wrapper,
+            payload=json.loads(wrapper["payload"]),
+        )
+        checkin_mapping[item.processing_key].append(item)
+
+    # Submit check-in groups for processing
+    with sentry_sdk.start_transaction(op="process_batch", name="monitors.monitor_consumer"):
+        futures = [
+            _checkin_workwer.submit(process_checkin_group, group)
+            for group in checkin_mapping.values()
```

Comment:
I had some additional thoughts since our conversation on Slack.  If you want to improve utilization we can sort by the number of checkins per monitor (in descending order -- big monitors before small monitors).  So something like this:

```python
sorted(checkin_mapping.values(), key=lambda n: len(n), reverse=True)
```

If the opposite were true then you might process the big monitor after all of the small monitors have completed.  Leaving you with a bunch of idle threads.

There may be more clever ways of doing this but my thought is since the runtime of each checkin is unknown (checkins require reaching out to a customer's network correct?) you can't really know how to weight each monitor.  Sorting by number of checkins is a sort of best effort approach to solving this.

Reply:
I considered this as well, but I'm not too sure we'll ever be dealing with too many check-ins per monitor.

It's much more likely the workload here is going to be for the most part processing many monitors that just have single check-ins.

> checkins require reaching out to a customer's network correct

No this is not true, it's not up-time monitoring, a check-in just represents a customer telling us their job has started or completed. The runtime for processing is known.

Reply:
We can probably track a metric for this -- Within a batch how many monitors have multiple check-ins, and what is the average